pkg/reconciler/kubernetes/adapter.go

```diff
@@ -46,138 +65,412 @@ type Config struct {
 	RetryDelay       time.Duration
 }
 
+func (c *Config) validate() error {
+
+	if c.MaxRetries < 0 {
+		return fmt.Errorf("config MaxRetries cannot be < 0 (got %d)", c.MaxRetries)
+	}
+	if c.MaxRetries == 0 {
+		c.MaxRetries = maxRetries
+	}
+	if c.RetryDelay < 0 {
+		return fmt.Errorf("config RetryDelay cannot be < 0 (got %d)", c.RetryDelay)
+	}
+	if c.RetryDelay == 0 {
+		c.RetryDelay = retryDelay
+	}
+
+	if c.ProgressInterval < 0 {
+		return fmt.Errorf("config ProgressInterval cannot be < 0 (got %d)", c.ProgressInterval)
+	}
+	if c.ProgressInterval == 0 {
+		c.ProgressInterval = progressTrackerInterval
+	}
+
+	if c.ProgressTimeout < 0 {
+		return fmt.Errorf("config ProgressTimeout cannot be < 0 (got %d)", c.ProgressTimeout)
+	}
+	if c.ProgressTimeout == 0 {
+		c.ProgressTimeout = progressTrackerTimeout
+	}
```

Comment:
This is pretty hard to understand, I would suggest having a `switch` statement first to deal with all errors and after that just assign values when needed:
```suggestion
        switch {
            case c.MaxRetries < 0:
		return fmt.Errorf("config MaxRetries cannot be < 0 (got %d)", c.MaxRetries)
	    case c.RetryDelay < 0:
		return fmt.Errorf("config RetryDelay cannot be < 0 (got %d)", c.RetryDelay)
	    case c.ProgressInterval < 0:
		return fmt.Errorf("config ProgressInterval cannot be < 0 (got %d)", c.ProgressInterval)
	    case c.ProgressTimeout < 0:
		return fmt.Errorf("config ProgressTimeout cannot be < 0 (got %d)", c.ProgressTimeout)
        }
        
	if c.MaxRetries == 0 {
		c.MaxRetries = maxRetries
	}
	if c.RetryDelay == 0 {
		c.RetryDelay = retryDelay
	}
	if c.ProgressInterval == 0 {
		c.ProgressInterval = progressTrackerInterval
	}
	if c.ProgressTimeout == 0 {
		c.ProgressTimeout = progressTrackerTimeout
	}
```

Reply:
nice, it's much clear! I actually extracted this code to a dedicate file named `adapterconfig.go`, will take you change but skip this commit suggestion.

Reply:
btw, we have more `validate()` code with the previous pattern (I just copied from there:)), we can do the refactoring to them later.

Reply:
yes, saw the change! I will resolve this and keep an eye for other occurrences.

It also has the advantage of not assigning values without need in case there is an error further below.